### DISCUSSION TITLE
SSH-integration

### DISCUSSION THREAD

**Discussion by Arian-D** [ghostty-org/ghostty#3087] [Feature Requests, Ideas]
Hi!
I'm a former Wezterm user, and the main reason I switched over to Ghostty after trying it was stability and native GTK look. The one thing I miss after about a month or so of usage is ssh-integration.

In Wezterm, the user is able to launch new tabs or other forms of sessions in a remote host, and subsequent sessions launched from there are also within the remote host.

I don't know the logistics, design decisions, and implementation work required to do this for Ghostty, but I thought I'd bring it up.

There's https://github.com/ghostty-org/ghostty/discussions/2353 for an API to and https://github.com/ghostty-org/ghostty/pull/1701 for a server model (which I personally prefer for keeping persistent remote sessions), but I'm curious to know if this is a planned feature or not on the roadmap.

Thanks!

**Comment by blychs** [comment]
+1 for this. I am a lover of Kitty, and have also been enjoying Ghostty quite a bit these past two days (thanks and congrats for the great work!). I will probably be using both back and forth in the future. Both seem great! 
Of all the features of Kitty, this is probably the one I miss the most in Ghostty). I don't care too much about remote persistence, since setting the simple layout I use on login is not too hard and (hopefully) I don't need to interrupt processes midway too often. However, being able to open new panes or tabs on the same connection would make the experience a lot more pleasant, and allow me to ditch using tmux too much.

**Comment by koramangalacode** [comment]
I know this isn't a builtin solution but here's a workaround that allows you to assign an ssh command to a window and its future tabs

Create directories that map to ssh hosts
```sh
mkdir ~/ssh-into/web-server
mkdir ~/ssh-into/database-server
mkdir ~/ssh-into/mail-server
```

Add a function to your shell that will ssh into your servers
```sh
# ~/.zshrc

ssh-into-server() {
  PWD=$HOME/ssh-into/$1
  ssh $1
}

[[ $PWD = $HOME/ssh-into/web-server ]] && ssh-into-server web-server
[[ $PWD = $HOME/ssh-into/database-server ]] && ssh-into-server database-server
[[ $PWD = $HOME/ssh-into/mail-server ]] && ssh-into-server mail-server
```

Run `ssh-into-server web-server` in the first tab of your new window. New tabs will load with the ssh session.

There may be other ways to do this but `$PWD` is passed on to the next tab and it's easily configurable.

**Reply by dgl** [reply]
I love this idea, but it is a bit manual, I don't like having to use a different command, here's an alternative adjusted from something I've been doing with my setup:

In `.ssh/config`:

```
Match Host *
  PermitLocalCommand yes
  LocalCommand ssh-pwd-update %h
```

Add ~/bin/ssh-pwd-update and chmod +x it:

```shell
#!/bin/sh
host="${1:?For usage from ssh_config only}"
mkdir -p "${HOME}/.ssh-pwd/${host}"
printf '\e]7;%s\a' "file://localhost${HOME}/.ssh-pwd/${host}" >> /dev/tty
```

Then in .zshrc (should work in bash too):

```shell
[[ ${PWD/$HOME\/.ssh-pwd\/} != ${PWD} ]] && exec ssh "$(basename "${PWD}")"
```

Note a cool aspect of this on macOS is if you customise the icon for ~/.ssh-pwd/host in Finder it will show that icon as the proxy icon when SSHed into that host.

Personally I think a neat solution for ghostty itself which isn't hacky like this could be extending OSC 7 to support hosts other than localhost in the `file://` URL, that way it would be possible to have directory tracking on remote systems too. (This also overlaps with a nice way to install the shell integration on remote servers.)

It would also be useful to have different key sequence to open a tab / window without transfering the state (i.e. there's probably another feature request there too, to have a keybind action to open a tab / window while ignoring the session state).

**Reply by koramangalacode** [reply]
Cool tip about the proxy icon!

**Reply by dgl** [reply]
Without a key combo to open a window / tab ignoring the state it gets a bit annoying, after trying it for a bit I've changed the command to:

```
if [[ ${PWD/$HOME\/.ssh-pwd\/} != ${PWD} ]]; then
  echo -n "SSH to $(basename "${PWD}")? " && read c && [[ ${c:0:1} != n ]] && exec ssh "$(basename "${PWD}")" || cd
fi
```

i.e. you can say n to not establish an SSH session, hit enter otherwise.

**Comment by italogv** [comment]
Other kind of SSH integration I would like to see is something in the idea of [Kitty Remote Files](https://sw.kovidgoyal.net/kitty/kittens/remote_file/), where you can "easily Edit, Open or Download files from a computer into which you are SSHed".
Kitty does this using OSC 8 (https://github.com/ghostty-org/ghostty/pull/1928).

**Comment by DAmesberger** [comment]
if you are interested, I implemented this (hand-designed but mostly vibe-coded, I use it as my daily driver): https://github.com/DAmesberger/ghostty/tree/feature/ssh-remote-sessions

You can read a bit of docs here: https://github.com/DAmesberger/ghostty/blob/feature/ssh-remote-sessions/SSH_REMOTE_SESSIONS.md

It implements a ghostty-daemon for the remote which is auto-deployed. you can create a new remote ssh session in this ghostty-fork with newly added ghostty commands, detach from them and re-attach later. it has named sessions (similar to tmux/zellij) and a session can have multiple tabs/splits. I also added color-coded tabs per session, so you can see easily which tabs belong to which session. A session manager is also present which lets you delete sessions etc. If there is any interest in it I might devote a bit more time to it to polish it.
